tools/canary_audit/workspace.py

```
@@ -0,0 +1,252 @@
+	normalized = safe_relative_path(relative)
+	_reject_symlink_components(root, normalized, include_leaf=reject_leaf_symlink)
+	candidate = (root / PurePosixPath(normalized)).resolve(strict=must_exist)
+	if not _is_within(candidate, root):
+		raise WorkspaceError(f"path escapes repository workspace: {relative!r}")
+	return candidate
+
+
+def _git_file_names(root: Path) -> list[str] | None:
+	try:
+		result = subprocess.run(
+			["git", "ls-files", "--cached", "--others", "--exclude-standard", "-z"],
```

Ruff (0.15.20)

[error] Starting a process with a partial executable path

(S607)